rust/xaynet-server/src/storage/s3.rs

```diff
@@ -0,0 +1,289 @@
+    /// };
+    ///
+    /// let s3_settings = S3Settings {
+    ///     region,
+    ///     access_key: String::from("minio"),
+    ///     secret_access_key: String::from("minio123"),
+    ///     buckets: S3BucketsSettings {
+    ///         global_models: String::from("global-models"),
+    ///     },
+    /// };
+    ///
+    /// let store = Client::new(s3_settings);
+    /// ```
+    pub fn new(settings: S3Settings) -> S3Result<Self> {
+        let credentials_provider =
+            StaticProvider::new_minimal(settings.access_key, settings.secret_access_key);
+
+        let dispatcher = HttpClient::new()?;
+        Ok(Self {
+            buckets: Arc::new(settings.buckets),
+            s3_client: S3Client::new_with(dispatcher, credentials_provider, settings.region),
+        })
+    }
+
+    /// Uploads a global model.
+    pub async fn upload_global_model(&self, key: &str, global_model: &Model) -> S3Result<()> {
+        // As key for the global model we use the seed of the round in which the global model was
+        // created. We also might store the round number for easier debugging?
+        debug!("store global model: {}", key);
+        let data = bincode::serialize(global_model)?;
+        self.upload(&self.buckets.global_models, key, data)
+            .await
+            .map_err(From::from)
+            .map(|_| ())
+    }
+
+    /// Creates the `global_models` bucket.
+    pub async fn create_global_models_bucket(&self) -> S3Result<()> {
+        debug!("create global-models bucket");
+        match self.create_bucket("global-models").await {
+            Ok(_)
+            | Err(RusotoError::Service(CreateBucketError::BucketAlreadyExists(_)))
+            | Err(RusotoError::Service(CreateBucketError::BucketAlreadyOwnedByYou(_))) => Ok(()),
+            Err(err) => Err(S3Error::from(err)),
+        }
+    }
+
+    // Uploads an object to the given bucket.
+    async fn upload(
+        &self,
+        bucket: &str,
+        key: &str,
+        data: Vec<u8>,
+    ) -> Result<PutObjectOutput, RusotoError<PutObjectError>> {
+        let req = PutObjectRequest {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+            body: Some(StreamingBody::from(data)),
+            ..Default::default()
+        };
+        self.s3_client.put_object(req).await
+    }
+
+    // Creates a new bucket with the given bucket name.
+    async fn create_bucket(
+        &self,
+        bucket: &str,
+    ) -> Result<CreateBucketOutput, RusotoError<CreateBucketError>> {
+        let req = CreateBucketRequest {
+            bucket: bucket.to_string(),
+            ..Default::default()
+        };
+        self.s3_client.create_bucket(req).await
+    }
+}
+
+#[cfg(test)]
+use rusoto_s3::{
+    Delete,
+    DeleteObjectsOutput,
+    DeleteObjectsRequest,
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    ObjectIdentifier,
+};
+
+#[cfg(test)]
+impl Client {
```

Comment:
this impl could be moved into the `tests` module below. but if it is also needed for other tests elsewhere it probably  should be moved to some test helper module.

Reply:
I agree but I would like to keep it here to make adding the feature flag easier in #567 .
this way I only need to add the flag to `mod  s3` in `rust/xaynet-server/src/storage/mod.rs`

Reply:
i'm not sure if i get this, because adding the flag to the `s3` module will also enable/disable the submodules which includes `s3::tests`. but if it's easier this way, then i'm ok with that.

Reply:
ah sry my bad, I thought the idea was to add this to `rust/xaynet-server/src/storage/tests/mod.rs`
That make sense I can do that👍